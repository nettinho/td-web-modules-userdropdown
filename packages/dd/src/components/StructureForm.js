import _ from "lodash/fp";
import React from "react";
import PropTypes from "prop-types";
import { bindActionCreators } from "redux";
import { connect } from "react-redux";
import { Form } from "semantic-ui-react";
import { updateStructure } from "../routines";

export class StructureForm extends React.Component {
  static propTypes = {
    structure: PropTypes.object,
    structureUpdating: PropTypes.bool,
    updateStructure: PropTypes.func
  };

  constructor(props) {
    super(props);
    const { structure } = props;
    this.state = { structure, edit: false };
    this.handleEdit = this.handleEdit.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.handleKeyDown = this.handleKeyDown.bind(this);
  }

  componentDidUpdate(prevProps) {
    if (prevProps.structureUpdating && !this.props.structureUpdating) {
      this.setState({ edit: false, structure: this.props.structure });
    }
  }

  handleKeyDown(e) {
    if (e.key === "Escape") {
      this.setState({ edit: false, structure: this.props.structure });
    }
  }

  handleEdit(e) {
    this.setState({ edit: true });
    e.preventDefault();
  }

  handleChange(e, data) {
    e.preventDefault();
    const { name, value } = data;
    const { structure } = this.state;
    const nextStructure = Object.assign({}, structure, { [name]: value });
    this.setState({ structure: nextStructure });
  }

  handleSubmit() {
    const { structure } = this.state;
    this.props.updateStructure({ data_structure: structure });
  }

  render() {
    const { structureUpdating } = this.props;
    const { structure, edit } = this.state;
    if (edit) {
      return (
        <Form size="small">
          <Form.Group>
            <Form.Input
              autoComplete="off"
              name="description"
              value={structure.description}
              onKeyDown={this.handleKeyDown}
              onChange={this.handleChange}
              required
              inline
              focus
              action={{
                icon: "check",
                onClick: this.handleSubmit,
                loading: structureUpdating,
                disabled: structureUpdating || !structure.description
              }}
            />
          </Form.Group>
        </Form>
      );
    } else {
      return (
        <p onClick={this.handleEdit}>
          {structure.description || "No definition"}
        </p>
      );
    }
  }
}

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators({ updateStructure }, dispatch)
});

const mapStateToProps = ({ structure, structureUpdating }) => ({
  structure: _.pick(["id", "description"])(structure),
  structureUpdating
});

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(StructureForm);
