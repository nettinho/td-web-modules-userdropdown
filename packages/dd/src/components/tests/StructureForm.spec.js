import React from "react";
import { shallow } from "enzyme";
import { StructureForm } from "../StructureForm";

const preventDefault = () => {};

const updateField = (wrapper, instance, name, value) =>
  wrapper
    .find(instance)
    .simulate("change", { preventDefault }, { name, value });

describe("<StructureForm />", () => {
  const updateStructure = jest.fn();
  const structure = {
    id: 1,
    name: "s1",
    description: "dd",
    group: "gp",
    system: "sys"
  };
  const props = { structure, structureUpdating: false, updateStructure };

  it("matches the latest snapshot", () => {
    const wrapper = shallow(<StructureForm {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it("matches the latest snapshot (editing)", () => {
    const wrapper = shallow(<StructureForm {...props} />);
    wrapper.find("p").simulate("click", { preventDefault });
    expect(wrapper).toMatchSnapshot();
  });

  it("becomes editable when the description is clicked", () => {
    const wrapper = shallow(<StructureForm {...props} />);
    expect(wrapper.find("p").length).toEqual(1);
    expect(wrapper.find("Form").length).toEqual(0);
    wrapper.find("p").simulate("click", { preventDefault });
    expect(wrapper.find("p").length).toEqual(0);
    expect(wrapper.find("Form").length).toEqual(1);
  });

  it("updates the state and description field value on change", () => {
    const wrapper = shallow(<StructureForm {...props} />);
    const newValue = "Something else";
    wrapper.find("p").simulate("click", { preventDefault });
    expect(wrapper.find("Form").length).toBe(1);
    expect(wrapper.find("FormInput").length).toBe(1);
    expect(wrapper.find("FormInput").props().value).toBe("dd");
    updateField(wrapper, "FormInput", "description", newValue);
    expect(wrapper.state("structure").description).toEqual(newValue);
    expect(wrapper.find("FormInput").props().value).toEqual(newValue);
  });
});
