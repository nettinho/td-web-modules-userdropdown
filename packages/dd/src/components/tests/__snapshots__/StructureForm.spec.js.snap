// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StructureForm /> matches the latest snapshot (editing) 1`] = `
<Form
  as="form"
  size="small"
>
  <FormGroup>
    <FormInput
      action={
        Object {
          "disabled": false,
          "icon": "check",
          "loading": false,
          "onClick": [Function],
        }
      }
      as={[Function]}
      autoComplete="off"
      control={[Function]}
      focus={true}
      inline={true}
      name="description"
      onChange={[Function]}
      onKeyDown={[Function]}
      required={true}
      value="dd"
    />
  </FormGroup>
</Form>
`;

exports[`<StructureForm /> matches the latest snapshot 1`] = `
<p
  onClick={[Function]}
>
  dd
</p>
`;
